blog/2024-05-30.md

```diff
@@ -0,0 +1,123 @@
+
+![waiting on a loop](../website/static/docs/blog/legendary.jpg)
+
+<!-- truncate -->
+
+## The problem of waiting
+Currently, smart contracts have no way to wait for an external event to happen. This can be a problem when the contract relies on an external service to provide a result.
+
+We encountered this issue while implementing [Chain Signatures](/concepts/abstraction/chain-signatures), which work by requiring an external service to provide a signature.
+
+Until now, the only workaround has been to make the contract call itself in a loop, checking on each iteration if the result is ready. Each call delays the result by one block (~1 second), allowing the contract to wait almost a minute before running out of gas.
+
+![waiting on a loop](../website/static/docs/blog/contract-wait-loop.png)
+*Until now, contracts had to wait by calling themselves until a external service replies... more often than not the contract will run out of gas waiting*
+
+While this method works, it's far from ideal. It wastes a lot of gas on looping and - more often than not - runs out of gas, forcing the user to retry the transaction.
+
+## Yield and Resume
+Starting from version `1.40` of the protocol, developers can **delay the execution** of a function until certain conditions are met (e.g. an external service provides a result).
+
+This way, instead of the contract calling itself on a loop waiting, the contract can simply **yield** calling the function that gives the result. When an external response is provided, the contract will **resume** and return the result.
+
+![waiting on a yield](../website/static/docs/blog/contract-wait-yield.png)
+*Contracts can now yield the execution of a function until an external service signals that the result is ready*
+
+### What is exactly being yielded?
+It is important to notice that the contract is not **halting** or **blocking** its ability to execute, nor **halting in the middle of a function** to later resume it. 
+
+In the same way that a function can return a promise to call another contract, now a function can return a **yield** to call another function.
+
+Indeed, the contract is not halting, but simply **delaying the execution of a callback** until an external agent signals that it is ok to resume.
+
+People can keep calling functions on the contract between a `yield`/`resume`, and the function that creates the `yield` can be called multiple times.
+
+:::tip
+The state **can change** between the `yield` and the `resume`, since people can keep interacting with the contract.
+
+Moreover, since the function used to signal is public, developers must make sure to guard it properly to avoid unwanted calls. This can be done by simply checking the caller of the function.
+:::
+
+### How does it change for the user?
+Between the `yield` and `resume` the user will simply be waiting to receive the result. The protocol will wait 200 blocks - around 4 minutes - after which it will resume the yielded method passing it a timeout.
```

Comment:
nit: This would feel more precise if we can somehow express it conditionally; the timeout is only delivered if a user-submitted resume call is not made within the 200 blocks.